core/arch/arm/plat-qcom/q6dsp.c

```diff
@@ -0,0 +1,165 @@
+	.pas_id = PAS_ID_WPSS,
+	.base.pa = WPSS_BASE,
+};
+#endif
+
+static void kodiak_wpss_start(struct thread_smc_args *args)
+{
+	uint32_t val;
+	uint64_t timer;
+	vaddr_t base = io_pa_or_va(&wpss_dsp_data.base, 0x500);
+	/* Enable the XO clock */
+	io_write32(base + QDSP6SS_XO_CBCR, 1);
+
+	/* Enable the QDSP6SS sleep clock */
+	io_write32(base + QDSP6SS_SLEEP_CBCR, 1);
+
+	/* Enable the QDSP6 core clock */
+	io_write32(base + QDSP6SS_CORE_CBCR, 1);
+
+	/* Program boot address */
+	io_write32(base + RST_EVB_REG, wpss_dsp_data.firmware_base >> 4);
+
+	/* Flush configuration */
+	dsb();
+
+	/* De-assert QDSP6 stop core. QDSP6 will execute after out of reset */
+	io_write32(base + CORE_START_REG, LPASS_BOOT_CORE_START);
+
+	/* Trigger boot FSM to start QDSP6 */
+	io_write32(base + BOOT_CMD_REG, LPASS_BOOT_CMD_START);
+
+	/* Wait for core to come out of reset */
+	timer = timeout_init_us(BOOT_FSM_TIMEOUT);
+	do {
+		val = io_read32(base + BOOT_STATUS_REG);
+		if ((val & BIT(0)) != 0)
+			break;
+		if (timeout_elapsed(timer))
+			break;
+		udelay(10);
+	} while (0);
+
+	if ((val & BIT(0)) != 0) {
+		EMSG("Timed out waiting for WPSS to boot :(");
+		args->a0 = -4;
+		return;
+	}
+
+	args->a0 = OPTEE_SMC_RETURN_OK;
+	args->a2 = 0;
+	args->a3 = 0;
+	args->a4 = 0;
+	args->a5 = 0;
+}
+
+#define QCOM_SCM_PIL_PAS_INIT_IMAGE	0x0201
+#define QCOM_SCM_PIL_PAS_MEM_SETUP	0x0202
+#define QCOM_SCM_PIL_PAS_AUTH_AND_RESET	0x0205
+#define QCOM_SCM_PIL_PAS_SHUTDOWN	0x0206
+#define QCOM_SCM_PIL_PAS_IS_SUPPORTED	0x0207
+#define QCOM_SCM_PIL_PAS_MSS_RESET	0x020a
+
+static inline const char *smc_decode_pil_cmd(uint32_t cmd)
+{
+	switch (cmd) {
+	case QCOM_SCM_PIL_PAS_INIT_IMAGE: return "init_image";
+	case QCOM_SCM_PIL_PAS_MEM_SETUP: return "mem_setup";
+	case QCOM_SCM_PIL_PAS_AUTH_AND_RESET: return "auth_and_reset";
+	case QCOM_SCM_PIL_PAS_SHUTDOWN: return "shutdown";
+	case QCOM_SCM_PIL_PAS_IS_SUPPORTED: return "is_supported";
+	case QCOM_SCM_PIL_PAS_MSS_RESET: return "mss_reset";
+	default:  return "???";
+	}
+}
+
+static inline const char *pas_id_name(uint32_t pas_id)
+{
+	switch (pas_id) {
+	case PAS_ID_WPSS: return "wpss";
+	default: return "???";
+	}
+}
+
+void qcom_handle_pil_smc(struct thread_smc_args *args)
+{
+	uint32_t sv_cmd = args->a0 & 0xffff;
+
+	DMSG("Handling %s cmd for %s", smc_decode_pil_cmd(sv_cmd),
+	     pas_id_name(args->a2));
+
+	if (args->a2 != PAS_ID_WPSS) {
+		EMSG("Ignoring request for unsupported DSP 0x%lx", args->a2);
+		args->a0 = OPTEE_SMC_RETURN_EBADCMD;
+		return;
+	}
+
+	switch (sv_cmd) {
+	case QCOM_SCM_PIL_PAS_INIT_IMAGE:
+		/* FIXME: handle MBN parsing stuff here */
+		break;
+	case QCOM_SCM_PIL_PAS_MEM_SETUP:
+		/* Save base address of firmware */
+		wpss_dsp_data.firmware_base = args->a3;
+		break;
+	case QCOM_SCM_PIL_PAS_AUTH_AND_RESET:
+		qcom_clock_enable(QCOM_CLKS_WPSS);
+		kodiak_wpss_start(args);
+		break;
+	case QCOM_SCM_PIL_PAS_SHUTDOWN:
+	case QCOM_SCM_PIL_PAS_IS_SUPPORTED:
+	case QCOM_SCM_PIL_PAS_MSS_RESET:
+		break;
```

Comment:
Please make sure you don't handle these with an OK return status if they're not implemented. We have enough broken firmware versions already. :sweat_smile: 

- SHUTDOWN should return some "not implemented" error code.
- You can probably just remove the MSS_RESET one, I think this was pretty much just used on good old msm8916.

Keeping IS_SUPPORTED here is okay I guess (should just return OK), but maybe add a comment that it's only okay because the if statement further above already checks for `if (args->a2 != PAS_ID_WPSS)`.

Reply:
* SHUTDOWN now returns -ENOTAVAIL
* Dropped MSS_RESET and IS_SUPPORTED since they're specific to the old WCNSS driver anyways

And yeah cleaned up the logic so we're not misbehaving

Reply:
> Dropped MSS_RESET and IS_SUPPORTED since they're specific to the old WCNSS driver anyways

I don't think IS_SUPPORTED is specific to WCNSS, even though it's only used by that driver in upstream Linux at the moment. If I remember correctly, QC firmware still implements it nowadays.

This call would be useful to have if someone wanted to add runtime selection of PAS vs direct booting in Linux. Can you try implementing it? It should be straightforward, just add a case for it and set `a1 = true`. (Ideally test it by adding `qcom_scm_pas_supported(PAS_ID_WPSS)` somewhere in Linux to be sure...)

Reply:
OK I re-added the IS_SUPPORTED call and validated it from the PAS driver, even though it's unused it should work as expected if that changes.